### ISSUE TITLE
feat(arm64): Add support for arm64.

### ISSUE THREAD

**Issue by RedstoneWizard08** [revoltchat/self-hosted#37]
This is the issue for a PR I made (#35). There are more details there.

**Comment by VevooltDev** [comment]
Hi, @RedstoneWizard08!
We have successfully added support for arm64 instances. Please these instructions:
Here are the installation steps:
``` docker run --privileged --rm tonistiigi/binfmt --install amd64 ```
Just follow the Quick Start tutorial.

**Comment by RedstoneWizard08** [comment]
Yay!

**Comment by RedstoneWizard08** [comment]
Btw I have tried to use binfmt and qemu and emulation with several different methods, and they always end in errors, glitches, or general slowness. What I was talking about for this was more of a native solution that does not have to use emulation.

**Comment by Rexogamer** [comment]
That's more of a temporary solution than an actual fix (which we're interested in pursuing) - native ARM support is much better

**Comment by RedstoneWizard08** [comment]
Ok, good.

**Comment by eternalflamestudios** [comment]
> Hi, @RedstoneWizard08! We have successfully added support for arm64 instances. Please these instructions: Here are the installation steps: `docker run --privileged --rm tonistiigi/binfmt --install amd64` After that just follow the Quick Start tutorial.

This did not work for me. It still said something along the lines of "/Linux/ARM64 is not recognized as a supported platform"

**Comment by RedstoneWizard08** [comment]
Yeah. Binfmt/qemu is a temporary solution. Those arm64 images will be made available when arm64 binary builds are.

**Comment by fluxuat0r** [comment]
Is there any way for me to selfhost Revolt on an ARM64-based system like a Raspberry Pi in the meantime through emulation or something similar? The command that VevooltPro provided did not fix the issue.

**Comment by RioghanDer2te** [comment]
I tried hosting it on my Raspberry Pi 3.
<sub>turns out it has insufficient power, but I'd still like an opinion in case I upgrade to a RaPi 4 or 5</sub>
1. First tried it as mentioned in the quickstart
  -> No matching manifest for linux/arm64/v8
3. found a [fix](https://stackoverflow.com/a/73614527) that let me get started
  -> exec format error
5. undid step 2 and run the [emulation solution](https://github.com/revoltchat/self-hosted/issues/37#issuecomment-1234281914) from this issue
  -> MongoDB 5.0+ requires a CPU with AVX support

did my step 2 screw everything up or did an update make this impossible again?

**Comment by insertish** [comment]
Edited OP, this should only affect the web client now. Backend has arm64 support across all services.

**Comment by CornerSyrup** [comment]
Hi, I'm facing the same problem here.

By the "the web client" you mean "web" and "caddy"?
Can I just skip those 2 service and run the compose?

I'm still getting the `no matching manifest for linux/arm64/v8 in the manifest list entries`, perhaps it's something adjustable in deployment steps?

Thx for contributing and answering 🫰🏻

**Comment by Docteh** [comment]
I just tried this on a Raspberry Pi 4, I needed to run emulation for the web container (`platform: linux/amd64` in compose.yml), and for mongodb I needed to run a different container from https://github.com/themattman/mongodb-raspberrypi-docker

For the version 7.0.14 of mongodb that is provided by themattman I changed the healthcheck to go to mongo rather than mongosh

Oops, just noticed that I could have pinned and tried mongodb 4.4 as mentioned in the [self-hosted readme](https://github.com/revoltchat/self-hosted#mongo-compatibility)

**Comment by jktam** [comment]
Here's my experience trying to run this on an ARM64 instance (I'm using Oracle Cloud's Always Free instances):

Followed all the steps as shown in the README.md, but a few more changes to the compose.yml were necessary.

I looked up all the docker images whether in docker hub or the revoltchat github (just copy the path in the `image:` line). I saw whether I needed to change the image path or add a platform specification. I would either have to change the image path, for example, `docker.io/mongo` to `arm64v8/mongo`, or add `platform: arm64` OR `platform:arm64/v8` (check the tags).

I also commented out the services `web` and `crond`, which I assume breaks web client connectivity. These had no appropriate tags or supported images, it just wouldn't work for me.

Also found out I would need to build the desktop client for my specific instance ([https://github.com/orgs/revoltchat/discussions/846](https://github.com/orgs/revoltchat/discussions/846)), which I don't know if I'll get to.

Hope this helps some people getting here from search engines, as well as those looking for Discord alternatives after the CEO news.

TL;DR: I did not get this working on an ARM64 host, but hope the info here leads some people to success. I also hope some official support is achieved.

**Comment by Docteh** [comment]
@jktam crond according to the readme at https://github.com/revoltchat/backend is for cleaning up unused files, not required. But it is close to being available as an ARM64 container. I think I can just nudge it over the line in https://github.com/revoltchat/backend/pull/413

But if anyone needs the arm64 crond executable its available in `ghcr.io/revoltchat/base` container at `/home/rust/src/target/release/revolt-crond`. Building the Dockerfile at https://github.com/revoltchat/backend/blob/main/crates/daemons/crond/Dockerfile will pull down that base container, extract the one file and stick it in a container image for ya.

I did also learn that the binfmt stuff mentioned earlier doesn't stay after a reboot, so I might look into just building the web container locally and seeing where that gets me.

**Comment by LZeugirdor** [comment]
Mongo doesn't work on my pi 4, i understand the pi isn't particularly powerful but I would like to be able to set it up anyways. Is there any way to use postgres rather than mongo?

**Comment by Docteh** [comment]
@LZeugirdor did you try any of the alternatives for a mongo image in this thread? Switching a database engine is never trivial

**Comment by Khaztaroth** [comment]
Almost all the images work on my RaspberryPi 5, but I'm seeing an issue with "redis" `eqalpha/keydb` where it keeps reporting an out of memory error. Modifying `/boot/firmware/config.txt` to add `kernel=kernel8.img` makes that work but then the database starts reporting out of memory errors. Tried with `eqalpha/keydb:arm64_v6.3.4` but it didn't seem to change much.

Any other key database available that would work with this?

Edit: got it to work with `redis:8.2.4-alpine3.22` everything except voice chats seem to work as expected on my RaspberryPi 5 8gb

Edit: this only applies to Revolt not the proper Stoat web interface

**Comment by Dadadah** [comment]
> Almost all the images work on my RaspberryPi 5, but I'm seeing an issue with "redis" `eqalpha/keydb` where it keeps reporting an out of memory error. Modifying `/boot/firmware/config.txt` to add `kernel=kernel8.img` makes that work but then the database starts reporting out of memory errors. Tried with `eqalpha/keydb:arm64_v6.3.4` but it didn't seem to change much.
> 
> Any other key database available that would work with this?
> 
> Edit: got it to work with `redis:8.2.4-alpine3.22` everything except voice chats seem to work as expected on my RaspberryPi 5 8gb
> 
> Edit: this only applies to Revolt not the proper Stoat web interface

Once #209 is merged, the new versions should work, including the new web app and voice. Could you give it a shot? I think this issue can be closed but I want to make sure first.

**Comment by Khaztaroth** [comment]
> > Almost all the images work on my RaspberryPi 5, but I'm seeing an issue with "redis" `eqalpha/keydb` where it keeps reporting an out of memory error. Modifying `/boot/firmware/config.txt` to add `kernel=kernel8.img` makes that work but then the database starts reporting out of memory errors. Tried with `eqalpha/keydb:arm64_v6.3.4` but it didn't seem to change much.
> > Any other key database available that would work with this?
> > Edit: got it to work with `redis:8.2.4-alpine3.22` everything except voice chats seem to work as expected on my RaspberryPi 5 8gb
> > Edit: this only applies to Revolt not the proper Stoat web interface
> 
> Once [#209](https://github.com/stoatchat/self-hosted/pull/209) is merged, the new versions should work, including the new web app and voice. Could you give it a shot? I think this issue can be closed but I want to make sure first.

I've been using the latest docker instructions as they are posted with the one difference being the web image using the tag `addb6b7` instead of `6c5970f`

Everything works with the expected limitations that still are under development.

EDIT: I just realized I'm still using the `redis` that I posted on the previous comment, I will try the one from the docker instructions real quick

EDIT 2: Redis still needs to use `redis:8.2.4-alpine3.22` on my RaspberryPi 5 or it gets stuck with a "out of memory" error, everything else works fine.

**Comment by Dadadah** [comment]
We will be updating from keydb likely, so that should also address this. I will look at doing that to close this issue.

**Comment by LZeugirdor** [comment]
> [@LZeugirdor](https://github.com/LZeugirdor) did you try any of the alternatives for a mongo image in this thread? Switching a database engine is never trivial

I'm a simple person and honestly I feel like if I can't use docker pull it adds unnecessary extra steps, especially for updates. If these solutions were easily pullable for portainer I'd be open to it. But as of right now I will just opt not to host a stoat chat instance. It's frustrating since nearly every service I've seen is using mongo db and mongo db dropping support for older architectures feels unreasonable, at least to me.
